File: 2015/Day 11/day11.cpp
//
//  Advent of Code 2015
//  Day 11
//
//  By PatchesTheDipstick
//


//  System includes...


//  Third Party includes...



//  Project includes...

#include "../utils.hpp"

bool isValidPassword(std::string password)
{
	bool straight_count = false;
	for (int a = 0; a < password.length() - 2; a++)
	{
		if ( (password[a] + 1 == password[a + 1]) &&
		     (password[a + 1] + 1 == password[a + 2]))
		{
			straight_count = true;
		}
	}
	if (!straight_count)
	{
		return (false);
	}
	bool forbiden_letters = false;
	for (int a = 0; a < password.length(); a++)
	{
		if ((password[a] == 'i') || (password[a] == 'o') || (password[a] == 'l'))
		{
			forbiden_letters = true;
		}
	}
	if (forbiden_letters)
	{
		return (false);
	}
	bool pair_one = false, pair_two = false;
	for (int a = 0; a < password.length() - 1; a++)
	{
		if (password[a] == password[a + 1])
		{
			if (pair_one)
			{
				pair_two = true;
				break;
			}
			else
			{
				pair_one = true;
				a++;
			}
		}
	}
	if ( (!pair_one) || (!pair_two) )
	{
		return (false);
	}
	return (true);
}

void incrementCharacter(char& in_char)
{
	if (in_char == 'z')
	{
	    in_char = 'a';
	}
	else if ((in_char == 'h') || (in_char == 'n') || (in_char == 'k'))
	{
		in_char += 2;
	}
	else
	{
		in_char++;
	}

}

std::string nextPassword(std::string old_password)
{
	std::string return_value = old_password;
	
	do
	{
		for (int a = old_password.length() - 1; a >= 0; a--)
		{
			incrementCharacter(return_value[a]);
			if (return_value[a] != 'a')
				break;
		}
	} while (!isValidPassword(return_value));
	
	return (return_value);
}

int main(int argc, char** argv)
{
	
	std::string input = "hepxcrrq";
	
	char a = 'a', m = 'm', z = 'z';
	
	std::cout << a << ", " << m << ", " << z << ", ";
	incrementCharacter(a);
	incrementCharacter(m);
	incrementCharacter(z);
	std::cout << a << ", " << m << ", " << z << std::endl;
	std::string answer_1 = nextPassword(input);
	std::string answer_2 = nextPassword(answer_1);
	
	std::cout << "Advent of Code 2015 Day 11 answers:" << std::endl;
	std::cout << "    Part 1 : " << answer_1 << std::endl;
	std::cout << "    Part 2 : " << answer_2 << std::endl;

	return (0);

}

File: 2015/Day 11/day11.py
#
#  Advent of Code 2015
#  Day 
#
#  By PatchesTheDipstick
#

def increment_char(ch):
	if ch == "z":
		return "a"
	else:
		return chr(ord(ch) + 1)

def to_list(password):
	out_list = []
	for ch in password:
		out_list.append(ch)
	return out_list

def to_string(password):
	out_string = ""
	for ch in password:
		out_string += ch
	return out_string

def increment_password(password):
	carry = True
	index = len(password) - 1
	series = to_list(password)
	while carry:
		if password[index] != "z":
			series[index] = increment_char(series[index])
			carry = False
		else:
			series[index] = increment_char(series[index])
			index -= 1
	return to_string(series)

def is_valid(password):
	series = to_list(password)
	#  test first rule
	max_run_len = 0
	run_len = 0
	for i in range(len(series) - 1):
		if ord(series[i]) + 1 == ord(series[i + 1]):
			run_len += 1
		else:
			if run_len > max_run_len:
				max_run_len = run_len
			run_len = 1
	if max_run_len < 3:
		return False
	# test second rule
	if "i" in series or "o" in series or "l" in series:
		return False
	# test third rule
	doubles = []
	for i in range(len(password) - 1):
		if password[i] == password[i + 1]:
			doubles.append((i, i + 1))
	if len(doubles) < 2:
		return False
	else:
		new_doubles = []
		for i in range(len(doubles) - 1):
			if doubles[i][1] == doubles[i + 1][0]:
				doubles[i + 1] = (-1, -1)
	while (-1, -1) in doubles:
		doubles.remove((-1, -1))
	if len(doubles) < 2:
		return False
	return True



if __name__ == '__main__':
	
	input = "hepxcrrq"

	print("Advent of Code 2015 Day 03")
	
	current_password = input
	while not is_valid(current_password):
		current_password = increment_password(current_password)

	print("  Part one:", current_password)

	current_password = increment_password(current_password)
	while not is_valid(current_password):
		current_password = increment_password(current_password)

	print("  Part two:", current_password)	
	